### PR TITLE
[MM-30432]: Allow users to specify different desktop notification sounds per channel

### PR DESCRIPTION
**Pull request by kyeongsoosoo**
#### Summary

I added `desktop_sound` and `desktop_notification_sound` props to `ChannelNotifyProps` (https://github.com/mattermost/mattermost-webapp/pull/11605)
I modified the server code to make it work.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/21191
https://mattermost.atlassian.net/browse/MM-30432

#### Release Note

```release-note
Allow users to specify different desktop notification sounds per channel
```

### CODE REVIEW

**Comment by kyeongsoosoo** [comment]
@jgilliam17 I updated the branch. Is there anything else I need to do?

**Comment by jgilliam17** [comment]
Thanks @kyeongsoosoo I'll retest today and add few reviewers

**Comment by jgilliam17** [comment]
@noxer and @isacikgoz adding you as your names popped up as github suggestion. Please feel free to remove or reassign to someone else as you see fit - thanks 🙂

File: app/channel.go [hunk 1]
```diff
@@ -1271,6 +1271,14 @@ func (a *App) UpdateChannelMemberNotifyProps(c request.CTX, data map[string]stri
 		filteredProps[model.DesktopNotifyProp] = desktop
 	}
 
+	if desktop_sound, exists := data[model.DesktopSoundNotifyProp]; exists {
+		filteredProps[model.DesktopSoundNotifyProp] = desktop_sound
+	}
+
+	if desktop_notification_sound, exists := data["desktop_notification_sound"]; exists {
```
Comment (noxer) [inline app/channel.go:1278]: Please also add this constant to `model/user.go` and reference it here (like with the other properties).

**Review by noxer — CHANGES_REQUESTED** [review]
Very nice work @kyeongsoosoo, thank you for your contribution. I'd like to request a minor change. We typically put fixed strings like this in constants. Otherwise it looks good to me.

PS: Please excuse the delay.

**Comment by kyeongsoosoo** [comment]
@noxer Thanks for a review. I added a constants.

**Comment by jgilliam17** [comment]
Thanks @kyeongsoosoo and apologies for delay in getting to this PR
We still need a second dev review, I'll add a request today, but please expect the possibility that there may be a delay due to holidays.

**Comment by jgilliam17** [comment]
Tested and commented on related webapp PR.

**Review by noxer — APPROVED** [review]
LGTM, thank you very much.

**Review by isacikgoz — APPROVED** [review]

**Comment by isacikgoz** [comment]
/update-branch

Sorry for the late review, this looks good to me. I'll re-request a QA review from @jgilliam17 to let her decide whether we should merge or not :)

**Comment by M-ZubairAhmed** [comment]
@kyeongsoosoo can you please update the conflicts

**Comment by M-ZubairAhmed** [comment]
@devinbinnie would you like to give it a look?

**Review by devinbinnie — APPROVED** [review]

**Comment by M-ZubairAhmed** [comment]
@kyeongsoosoo no mattermost-webapp is now archived repository. Please bring in the webapp part of the changes in this PR itself from https://github.com/mattermost/mattermost-webapp/pull/11605

**Comment by kyeongsoosoo** [comment]
@M-ZubairAhmed I moved the webapp code to this PR

**Comment by M-ZubairAhmed** [comment]
@kyeongsoosoo sorry it looks like a lot of conflicts got in the way :(

**Comment by kyeongsoosoo** [comment]
> @kyeongsoosoo sorry it looks like a lot of conflicts got in the way :(

I resolved the conflicts :)

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/sass/utils/_flex.scss [hunk 2]
```diff
@@ -4,6 +4,10 @@
     display: flex !important;
 }
 
+.justify-content-between {
```
[the remark this answers is not in the document]
Comment (kyeongsoosoo) [reply webapp/channels/src/sass/utils/_flex.scss:7]: I added a scss file and removed utility css.

**Review by M-ZubairAhmed — COMMENTED** [review]
The code looks good to me; I will defer to PM and subsequently UX review on this one. Thank you for getting this done :)

**Comment by kyeongsoosoo** [comment]
@M-ZubairAhmed I've experienced similar failures in other pr(#23117), would this be the same case?

**Comment by M-ZubairAhmed** [comment]
> @M-ZubairAhmed I've experienced similar server creation failures in other pr(#23117), would this be the same case?

Yes, it seems like that, let me try a different approach to get the server running

**Comment by M-ZubairAhmed** [comment]
I have created https://github.com/mattermost/mattermost-server/pull/23449 from repo's upstream branch, lets check if that will trigger a build

**Comment by M-ZubairAhmed** [comment]
Here the cloud URL for testing
https://mattermost-server-pr-23449.test.mattermost.cloud/

**Review by asaadmahmood — APPROVED** [review]
UX looks good @esethna and seems to be working fine too.

**Comment by esethna** [comment]
Moving PM review to @johndavidlugtu as this is in his teams AORs

**Review by johndavidlugtu — APPROVED** [review]
LGTM, thanks

File: webapp/channels/src/sass/utils/_flex.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply webapp/channels/src/sass/utils/_flex.scss:7]: This utility still exists here can you remove this?

File: webapp/channels/src/components/channel_notifications_modal/components/expand_view.tsx [hunk 3]
```diff
@@ -272,6 +312,72 @@ export default function ExpandView({
...
+                                checked={memberDesktopSound === DesktopSound.OFF}
+                                onChange={onChangeDesktopSound}
+                            />
+                            <FormattedMessage
+                                id='channel_notifications.sound.off.title'
+                                defaultMessage='Off'
+                            />
+                        </label>
+                    </div>
+                    {memberDesktopSound === DesktopSound.ON && (!isDesktopApp() || (window.desktop && semver.gte(window.desktop.version || '', '4.6.0'))) &&
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/channel_notifications_modal/components/expand_view.tsx:358]: why did we had to check for a specific desktop version here?

File: webapp/channels/src/components/channel_notifications_modal/components/expand_view.tsx [hunk 4]
```diff
@@ -1,41 +1,77 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
-import React, {ChangeEvent} from 'react';
+import React, {ChangeEvent, useMemo, useRef} from 'react';
 import {FormattedMessage} from 'react-intl';
 import {useSelector} from 'react-redux';
 
+import * as NotificationSounds from 'utils/notification_sounds';
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/channel_notifications_modal/components/expand_view.tsx:8]: can you selectively import here?

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]
This is looking amazing, i just have few nits before we can handover for QA

File: webapp/channels/src/components/channel_notifications_modal/components/expand_view.tsx (hunk 3 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/channel_notifications_modal/components/expand_view.tsx:358]: we have removed them https://github.com/mattermost/mattermost-server/pull/23558/files

Comment (kyeongsoosoo) [reply webapp/channels/src/components/channel_notifications_modal/components/expand_view.tsx]: I removed it :)

**Comment by kyeongsoosoo** [comment]
@M-ZubairAhmed Thanks for the review. I changed the codes.

**Review by M-ZubairAhmed — APPROVED** [review]
This looks good :shipit: Thank you

**Review by yasserfaraazkhan — APPROVED** [review]

**Pull request merged**